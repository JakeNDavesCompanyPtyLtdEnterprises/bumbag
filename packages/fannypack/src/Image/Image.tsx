import * as React from 'react';
import * as PropTypes from 'prop-types';
import classNames from 'classnames';
import { ImageProps as ReakitImageProps } from 'reakit/ts';

import _Image from './styled';

export type LocalImageProps = {
  children?: React.ReactNode;
  className?: string;
  /** How the image fits its bounds */
  fit?: 'cover' | 'contain';
  /** Positioning of the fitted image. Value can be "top", "left", "center", "right", "bottom" or an "x y" coordinate */
  fitPosition?: string;
  /** Fix the width of the image. It will not be responsive. */
  isFixed?: boolean;
};
export type ImageProps = LocalImageProps & ReakitImageProps;

export const Image: React.FunctionComponent<LocalImageProps> = ({ children, className, isFixed, ...props }) => (
  <_Image
    className={classNames({
      [className || '']: Boolean(className)
    })}
    isFixed={isFixed}
    {...props}
  >
    {children}
  </_Image>
);

Image.propTypes = {
  children: PropTypes.node,
  className: PropTypes.string,
  fit: PropTypes.oneOf(['cover', 'contain']) as PropTypes.Validator<LocalImageProps['fit']>,
  fitPosition: PropTypes.string,
  isFixed: PropTypes.bool
};
Image.defaultProps = {
  children: undefined,
  className: undefined,
  fit: undefined,
  fitPosition: undefined,
  isFixed: false
};

const C: React.FunctionComponent<ImageProps> = Image;
export default C;
