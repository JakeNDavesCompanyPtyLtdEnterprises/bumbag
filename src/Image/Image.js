// @flow
import React, { type Node } from 'react';
import classNames from 'classnames';

import _Image from './styled';

type Props = {
  use?: any,
  children: Node,
  className?: string,
  /** How the image fits its bounds */
  fit?: 'cover' | 'contain',
  /** Positioning of the fitted image. Value can be "top", "left", "center", "right", "bottom" or an "x y" coordinate */
  fitPosition?: string,
  /** Fix the width of the image. It will not be responsive. */
  isFixed?: boolean
};

const Image = ({ children, className, isFixed, ...props }: Props) => (
  <_Image
    className={classNames({
      [className || '']: Boolean(className)
    })}
    isFixed={isFixed}
    {...props}
  >
    {children}
  </_Image>
);

Image.defaultProps = {
  use: undefined,
  className: undefined,
  fit: undefined,
  fitPosition: undefined,
  isFixed: false
};

export default Image;
