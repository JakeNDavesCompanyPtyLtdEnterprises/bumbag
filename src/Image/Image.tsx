import * as React from 'react';
import classNames from 'classnames';
import { ImageProps as ReakitImageProps } from 'reakit/ts';

import _Image from './styled';

export interface LocalImageProps {
  use?: any;
  children?: React.ReactNode;
  className?: string;
  /** How the image fits its bounds */
  fit?: 'cover' | 'contain';
  /** Positioning of the fitted image. Value can be "top", "left", "center", "right", "bottom" or an "x y" coordinate */
  fitPosition?: string;
  /** Fix the width of the image. It will not be responsive. */
  isFixed?: boolean;
}
export type ImageProps = LocalImageProps & ReakitImageProps;

export const Image: React.FunctionComponent<LocalImageProps> = ({ children, className, isFixed, ...props }) => (
  <_Image
    className={classNames({
      [className || '']: Boolean(className)
    })}
    isFixed={isFixed}
    {...props}
  >
    {children}
  </_Image>
);

Image.defaultProps = {
  use: undefined,
  className: undefined,
  fit: undefined,
  fitPosition: undefined,
  isFixed: false
};

const C: React.FunctionComponent<ImageProps> = Image;
export default C;
