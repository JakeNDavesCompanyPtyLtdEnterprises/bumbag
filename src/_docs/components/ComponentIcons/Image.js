import React from 'react';

const SvgComponent = props => (
  <svg width={58} height={50} fill="none" {...props}>
    <path d="M57.143.389H0v48.897h57.143V.389z" fill="#444bc9" />
    <path
      fillRule="evenodd"
      clipRule="evenodd"
      d="M18.741 26.39l5.915 6.257 9.37-12.551 13.117 21.428H9.286L18.74 26.39z"
      fill="#fff"
    />
    <path d="M18.357 20a5.357 5.357 0 1 0 0-10.714 5.357 5.357 0 0 0 0 10.714z" fill="#fff" />
  </svg>
);

export default SvgComponent;
