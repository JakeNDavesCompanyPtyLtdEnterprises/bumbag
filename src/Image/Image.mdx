---
name: Image
route: /components/image
menu: Components
---

import { Playground, PropsTable } from 'docz';
import { Box } from '../primitives/index';
import Image from './index';

# Image

## Import

`import { Image } from 'fannypack'`

## Basic Usage

`<Image>` renders a `<img>` with `max-width: 100%`  and default reset styles applied.

<Playground>
  <Image src="https://raw.githubusercontent.com/bigbellies/fannypack/master/src/_docs/bean.jpg" alt="Bean" width={500} />
</Playground>

`<Image>` is also responsive by default, however, if you want to fix the width, add the `isFixed` prop.

<Playground>
  <Image isFixed src="https://raw.githubusercontent.com/bigbellies/fannypack/master/src/_docs/bean.jpg" alt="Bean" width={500} />
</Playground>

## Image fit

Setting a `fit` prop to `"contain"` means that the whole image will fit in the bounds and it will not lose its scale.

<Playground>
  <Image fit="contain" height={300} width={300} src="https://raw.githubusercontent.com/bigbellies/fannypack/master/src/_docs/bean.jpg" alt="Bean" backgroundColor="whitesmoke" />
</Playground>

Setting a `fit` prop to `"cover"` means that the image will fill the bounds and it will not lose its scale.

<Playground>
  <Image fit="cover" height={300} width={300} src="https://raw.githubusercontent.com/bigbellies/fannypack/master/src/_docs/bean.jpg" alt="Bean" backgroundColor="whitesmoke" />
</Playground>

### Positioning

You can change the position of a fitted image with the `fitPosition` prop. This prop mimics the `object-position` CSS attribute.

<Playground>
  <Image fit="contain" fitPosition="top" height={300} width={300} src="https://raw.githubusercontent.com/bigbellies/fannypack/master/src/_docs/bean.jpg" alt="Bean" backgroundColor="whitesmoke" />
</Playground>

<Playground>
  <Image fit="cover" fitPosition="10px 5px" height={300} width={300} src="https://raw.githubusercontent.com/bigbellies/fannypack/master/src/_docs/bean.jpg" alt="Bean" backgroundColor="whitesmoke" />
</Playground>

## Props

<PropsTable of={Image} />

## Theming

### Schema

```jsx
{
  base: string | Object,
  fixed: string | Object,
  fit: {
    base: string | Object,
    cover: string | Object,
    contain: string | Object
  }
}
```
