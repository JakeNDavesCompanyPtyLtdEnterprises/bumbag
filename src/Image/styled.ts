import { theme } from 'styled-tools';
import { Image } from 'reakit';

import styled, { css } from '../styled';
import { ImageProps } from './Image';

const fitProperties = (props: ImageProps) => {
  if (!props.fit) return;
  if (props.fit === 'contain') {
    return css`
      object-fit: contain;
      ${(props: ImageProps) =>
        props.fitPosition
          ? css`
              object-position: ${props.fitPosition};
            `
          : ''};

      ${theme('fannypack.Image.fit.base')};
      ${theme('fannypack.Image.fit.contain')};
    `;
  }
  if (props.fit === 'cover') {
    return css`
      object-fit: cover;
      ${(props: ImageProps) =>
        props.fitPosition
          ? css`
              object-position: ${props.fitPosition};
            `
          : ''};

      ${theme('fannypack.Image.fit.base')};
      ${theme('fannypack.Image.fit.cover')};
    `;
  }
  return null;
};
const fixedProperties = (props: ImageProps) => {
  if (!props.isFixed) return;
  return css`
    max-width: unset;

    ${theme('fannypack.Image.fixed')};
  `;
};

export default styled(Image)<ImageProps>`
  max-width: 100%;

  & {
    ${fitProperties};
  }
  & {
    ${fixedProperties};
  }

  ${theme('fannypack.Image.base')};
`;
