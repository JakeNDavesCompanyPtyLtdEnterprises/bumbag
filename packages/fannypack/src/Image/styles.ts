import { css, cssClass } from '../styled';
import { theme } from '../utils';

export const Image = styleProps => cssClass`
  max-width: 100%;

  ${styleProps.isFixed &&
    css`
      max-width: unset;

      ${theme(`${styleProps.themeKey}.css.fixed`)(styleProps)};
    `}

  ${styleProps.fit === 'contain' &&
    css`
      object-fit: contain;

      ${styleProps.fitPosition &&
        css`
          object-position: ${styleProps.fitPosition};
        `};

      ${theme(`${styleProps.themeKey}.css.contain`)(styleProps)};
    `}

  ${styleProps.fit === 'cover' &&
    css`
      object-fit: cover;

      ${styleProps.fitPosition &&
        css`
          object-position: ${styleProps.fitPosition};
        `};

      ${theme(`${styleProps.themeKey}.css.cover`)(styleProps)};
    `}

  & {
    ${theme(`${styleProps.themeKey}.css.root`)(styleProps)};
  }
`;
