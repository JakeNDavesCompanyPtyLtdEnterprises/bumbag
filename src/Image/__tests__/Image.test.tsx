import * as React from 'react';
import render from '../../_utils/tests/render';
import Image from '../Image';

it('renders correctly for a basic image', () => {
  const { container } = render(
    <Image src="https://raw.githubusercontent.com/fannypackui/fannypack/master/src/_docs/bean.jpg" alt="Bean" />
  );
  expect(container.firstChild).toMatchSnapshot();
});

it('renders correctly for a fixed image', () => {
  const { container } = render(
    <Image
      isFixed
      src="https://raw.githubusercontent.com/fannypackui/fannypack/master/src/_docs/bean.jpg"
      alt="Bean"
      width={500}
    />
  );
  expect(container.firstChild).toMatchSnapshot();
});

it('renders correctly for a fitted image (contain)', () => {
  const { container } = render(
    <Image
      fit="contain"
      height={300}
      width={300}
      src="https://raw.githubusercontent.com/fannypackui/fannypack/master/src/_docs/bean.jpg"
      alt="Bean"
    />
  );
  expect(container.firstChild).toMatchSnapshot();
});

it('renders correctly for a fitted image (cover)', () => {
  const { container } = render(
    <Image
      fit="cover"
      height={300}
      width={300}
      src="https://raw.githubusercontent.com/fannypackui/fannypack/master/src/_docs/bean.jpg"
      alt="Bean"
    />
  );
  expect(container.firstChild).toMatchSnapshot();
});

it('renders correctly for a fitted image with a position', () => {
  const { container } = render(
    <Image
      fit="contain"
      fitPosition="top"
      height={300}
      width={300}
      src="https://raw.githubusercontent.com/fannypackui/fannypack/master/src/_docs/bean.jpg"
      alt="Bean"
    />
  );
  expect(container.firstChild).toMatchSnapshot();
});

it('renders correctly for a fitted image with a position', () => {
  const { container } = render(
    <Image
      fit="contain"
      fitPosition="10px 5px"
      height={300}
      width={300}
      src="https://raw.githubusercontent.com/fannypackui/fannypack/master/src/_docs/bean.jpg"
      alt="Bean"
    />
  );
  expect(container.firstChild).toMatchSnapshot();
});
