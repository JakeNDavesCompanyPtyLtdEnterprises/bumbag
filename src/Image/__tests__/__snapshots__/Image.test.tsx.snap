// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly for a basic image 1`] = `
.c1 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
  background-color: unset;
  color: inherit;
}

.c1:focus:not(:focus-visible) {
  outline: none;
}

.c0 {
  max-width: 100%;
}

<img
  alt="Bean"
  class="c0 c1"
  src="https://raw.githubusercontent.com/fannypackui/fannypack/master/src/_docs/bean.jpg"
/>
`;

exports[`renders correctly for a fitted image (contain) 1`] = `
.c1 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
  background-color: unset;
  color: inherit;
}

.c1:focus:not(:focus-visible) {
  outline: none;
}

.c0 {
  max-width: 100%;
}

.c0 {
  object-fit: contain;
}

<img
  alt="Bean"
  class="c0 c1"
  height="300"
  src="https://raw.githubusercontent.com/fannypackui/fannypack/master/src/_docs/bean.jpg"
  style="height: 300px; width: 300px;"
  width="300"
/>
`;

exports[`renders correctly for a fitted image (cover) 1`] = `
.c1 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
  background-color: unset;
  color: inherit;
}

.c1:focus:not(:focus-visible) {
  outline: none;
}

.c0 {
  max-width: 100%;
}

.c0 {
  object-fit: cover;
}

<img
  alt="Bean"
  class="c0 c1"
  height="300"
  src="https://raw.githubusercontent.com/fannypackui/fannypack/master/src/_docs/bean.jpg"
  style="height: 300px; width: 300px;"
  width="300"
/>
`;

exports[`renders correctly for a fitted image with a position 1`] = `
.c1 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
  background-color: unset;
  color: inherit;
}

.c1:focus:not(:focus-visible) {
  outline: none;
}

.c0 {
  max-width: 100%;
}

.c0 {
  object-fit: contain;
  object-position: top;
}

<img
  alt="Bean"
  class="c0 c1"
  height="300"
  src="https://raw.githubusercontent.com/fannypackui/fannypack/master/src/_docs/bean.jpg"
  style="height: 300px; width: 300px;"
  width="300"
/>
`;

exports[`renders correctly for a fitted image with a position 2`] = `
.c1 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
  background-color: unset;
  color: inherit;
}

.c1:focus:not(:focus-visible) {
  outline: none;
}

.c0 {
  max-width: 100%;
}

.c0 {
  object-fit: contain;
  object-position: 10px 5px;
}

<img
  alt="Bean"
  class="c0 c1"
  height="300"
  src="https://raw.githubusercontent.com/fannypackui/fannypack/master/src/_docs/bean.jpg"
  style="height: 300px; width: 300px;"
  width="300"
/>
`;

exports[`renders correctly for a fixed image 1`] = `
.c1 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
  background-color: unset;
  color: inherit;
}

.c1:focus:not(:focus-visible) {
  outline: none;
}

.c0 {
  max-width: 100%;
}

.c0 {
  max-width: unset;
}

<img
  alt="Bean"
  class="c0 c1"
  src="https://raw.githubusercontent.com/fannypackui/fannypack/master/src/_docs/bean.jpg"
  style="width: 500px;"
  width="500"
/>
`;
